force_wfmanager/wfmanager_results_task.py

```diff
@@ -178,11 +190,224 @@ def _default_layout_default(self):
+            return False
+
+        current_file = dialog.path
+
+        if self._write_project_file(current_file):
+            self.current_file = current_file
+            return True
+
+        return False
+
+    def _write_project_file(self, file_path):
+        """ Writes a JSON file that contains the :attr:`Workflow` and
+        :attr:`AnalysisModel`.
+
+        """
+        try:
+            with open(file_path, 'w') as output:
+                # create a dictionary that contains analysis model,
+                # workflow and version that can be read back in by
+                # :class:`WorkflowReader`, and dump to JSON
+                project_json = {}
+                project_json['analysis_model'] = self.analysis_model.as_json()
+                project_json['workflow'] = WorkflowWriter() \
+                    .get_workflow_data(self.workflow_model)
+                project_json['version'] = WorkflowWriter().version
+                json.dump(project_json, output, indent=4)
+
+        except IOError as e:
+            error(
+                None,
+                'Cannot save in the requested file:\n\n{}'.format(
+                    str(e)),
+                'Error when saving the project'
+            )
+            log.exception('Error when saving Project')
+            return False
+
+        except Exception as e:
+            error(
+                None,
+                'Cannot save the Project:\n\n{}'.format(
+                    str(e)),
+                'Error when saving results'
+            )
+            log.exception('Error when the Project')
+            return False
+        else:
+            return True
+
+    def open_project(self):
+        """ Shows a dialog to open a JSON file and load the contents into
+        :attr:`Workflow` and :attr:`AnalysisModel`.
+
+        """
+
+        dialog = FileDialog(
+            action="open",
+            wildcard='JSON files (*.json)|*.json',
+        )
+
+        result = dialog.open()
+
+        if result is not OK:
+            return False
+
+        current_file = dialog.path
+
+        if self._load_project_file(current_file):
+            self.current_file = current_file
+            return True
+
+        return False
+
+    def _load_project_file(self, file_path):
+        """ Load contents of JSON file into:attr:`Workflow` and
+        :attr:`AnalysisModel`.
+
+        """
+
+        try:
+            with open(file_path, 'r') as fp:
+                project_json = json.load(fp)
+
+                # share the analysis model with the setup_task
+                self.analysis_model.from_dict(project_json['analysis_model'])
+                self.setup_task.analysis_model = self.analysis_model
+
+                # create two separate workflows, so that setup task can be
+                # edited without changing the results task copy
+                reader = WorkflowReader(self.setup_task.factory_registry)
+                fp.seek(0)
```

Comment:
Is this the part where it wouldn't copy things in a straightforward way?

Reply:
Yep, think this is the best way of doing it, rather than hacking __deepcopy__!